Updates

stefanganzer opened the pull request:
Hello Joe,

I recently watched one of your presentations where you use this project to explain different oauth2 flows. It helped me a lot to gain a better understanding of how the different parts play together.

This change set updates the demo to use the current Spring Boot version 2.7.3, and Authorization Server version 0.3.1.
The required changes to make the apps run were:
* having unique secrets
* using noop password hashing mechanism
* changing redirect uri from localhost to 127.0.0.1
* using builders where required
* using beans (as shown in Authorization Server documentation) instead of importing default configuration
* changing deprecated ClientAuthenticationMethod.BASIC to CLIENT_SECRET_BASIC

The only difference I noted when running the app is that the JWT ID (jti) seems not to be accessible/populated anymore. I did not dig deeper into this, as everything else was working.

Best regards
Stefan

jgrandja commented:
Hi @stefanganzer. Sorry for the delayed response. I've been quite busy with the major releases lately.

I haven't been staying up to date with this sample but I would be happy to merge this PR.

However, I would prefer to update this sample to use Spring Authorization Server 1.0 and Spring Boot 3.0.

Would you be able to update the PR?

stefanganzer commented:
Hi @jgrandja, no worries.
I have updated the PR to use the latest Spring Boot 3.0.1 and Authorization Server 1.0.0.

I have added one additional change that fails requests to services with an exception if the status is 3xx, instead of just returning `null`. Feel free to skip that commit - it was useful for debugging, though.

One issue I noted is that you get a not very informative error (`There was an unexpected error (type=None, status=999).`) after logging in, when you access the application with `http://localhost:8080` instead of `http://127.0.0.1:8080`. Only when you enable `trace` logging for `org.springframework.security.oauth2` you will find the cause:

```
TRACE 535415 --- [nio-9000-exec-3] .s.a.w.OAuth2AuthorizationEndpointFilter : Authorization request failed: [invalid_request] localhost is not allowed for the redirect_uri (http://localhost:8080/login/oauth2/code/login-client). Use the IP literal (127.0.0.1) instead.

org.springframework.security.oauth2.server.authorization.authentication.OAuth2AuthorizationCodeRequestAuthenticationException: localhost is not allowed for the redirect_uri (http://localhost:8080/login/oauth2/code/login-client). Use the IP literal (127.0.0.1) instead.
	at org.springframework.security.oauth2.server.authorization.authentication.OAuth2AuthorizationCodeRequestAuthenticationValidator.throwError(OAuth2AuthorizationCodeRequestAuthenticationValidator.java:208) ~[spring-security-oauth2-authorization-server-1.0.0.jar:1.0.0]
	at rg.springframework.security.oauth2.server.authorization.authentication.OAuth2AuthorizationCodeRequestAuthenticationValidator.validateRedirectUri(OAuth2AuthorizationCodeRequestAuthenticationValidator.java:114) ~[spring-security-oauth2-authorization-server-1.0.0.jar:1.0.0]
```
Maybe this error could be logged with a higher level by default?

jgrandja commented:
Thank you for your patience @stefanganzer. This is now merged!

I really appreciate you taking the time to update this sample 👍
